src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java

```diff
@@ -269,13 +271,252 @@ public Application2E2021(@Nonnull IMFCompositionPlaylistType imfCompositionPlayl
         }
     }
 
-    public static boolean isValidJ2KProfile(CompositionImageEssenceDescriptorModel imageDescriptor) {
+    private static boolean validateHT(CompositionImageEssenceDescriptorModel imageDescriptor,
+                                     IMFErrorLogger logger) {
+        boolean isValid = true;
+
+        J2KHeaderParameters p = imageDescriptor.getJ2KHeaderParameters();
+
+        if (p == null) {
+            logger.addError(
+                    IMFErrorLogger.IMFErrors.ErrorCodes.APPLICATION_COMPOSITION_ERROR,
+                    IMFErrorLogger.IMFErrors.ErrorLevels.FATAL,
+                    "APP2.HT: Missing or incomplete JPEG 2000 Sub-descriptor");
+            return false;
+        }
+
+        if (p.xosiz != 0 || p.yosiz != 0 || p.xtosiz != 0 || p.ytosiz != 0) {
+            logger.addError(
+                    IMFErrorLogger.IMFErrors.ErrorCodes.APPLICATION_COMPOSITION_ERROR,
+                    IMFErrorLogger.IMFErrors.ErrorLevels.FATAL,
+                    "APP2.HT: Invalid XOsiz, YOsiz, XTOsiz or YTOsiz");
+            isValid = false;
+        }
+
+        if (p.xtsiz < p.xsiz || p.ytsiz < p.ysiz) {
+            logger.addError(
+                    IMFErrorLogger.IMFErrors.ErrorCodes.APPLICATION_COMPOSITION_ERROR,
+                    IMFErrorLogger.IMFErrors.ErrorLevels.FATAL,
+                    "APP2.HT: Invalid XTsiz or XYsiz");
+            isValid = false;
+        }
+
+        /* components constraints */
+
+        if (p.csiz.length <= 0 || p.csiz.length > 4) {
+            logger.addError(
+                    IMFErrorLogger.IMFErrors.ErrorCodes.APPLICATION_COMPOSITION_ERROR,
+                    IMFErrorLogger.IMFErrors.ErrorLevels.FATAL,
+                    String.format("APP2.HT: Invalid number (%d) of components", p.csiz.length));
+            isValid = false;
+        }
+
+        if (p.csiz[0].ssiz > 15 || p.csiz[0].ssiz < 7) {
+            logger.addError(
+                    IMFErrorLogger.IMFErrors.ErrorCodes.APPLICATION_COMPOSITION_ERROR,
+                    IMFErrorLogger.IMFErrors.ErrorLevels.FATAL,
+                    String.format("APP2.HT: Invalid bit depth (%d)", p.csiz[0].ssiz + 1));
+            isValid = false;
+        }
+
+        for (int i = 0; i < p.csiz.length; i++) {
+            if (p.csiz[i].yrsiz != 1) {
+                logger.addError(
+                        IMFErrorLogger.IMFErrors.ErrorCodes.APPLICATION_COMPOSITION_ERROR,
+                        IMFErrorLogger.IMFErrors.ErrorLevels.FATAL,
+                        String.format("APP2.HT: invalid vertical sub-sampling for component %d", i));
+                isValid = false;
+            }
+            if (p.csiz[i].xrsiz != 1) {
+                if (i == 2 && p.csiz[i].xrsiz != 2) {
```

Comment:
I think this _may_ be a possible off-by-one bug?

If I am correctly interpreting the [2067-21 spec](https://pub.smpte.org/doc/st2067-21/20221124-pub/), I believe when the HT-J2K constraints (Annex I) refer to a specific component, these component numbers are **one-based**, rather than zero-based index values used by Java.

So in the case of the sub-sampling constraint, ` (Xrsizi=2 for i = {2,3} and XRsizi=1 for other i) `, I would interpret that as describing the _"second and third component"_. In the java code, that means you would be looking at csiz index where `i is 1 or 2`, instead of when `i is 2 or 3`

Reply:
![Screenshot 2024-06-25 at 4 09 13 PM](https://github.com/Netflix/photon/assets/45833563/f03ab807-c0a0-4c68-b560-bcf093dbdc19)

Reply:
> these component numbers are one-based, rather than zero-based index values used by Java.

Will fix.

Reply:
Addressed at https://github.com/Netflix/photon/pull/368/commits/c057b3eabdd43332629915f539f6c1f87d51fbc3